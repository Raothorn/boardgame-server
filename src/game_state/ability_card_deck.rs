use serde::Serialize;

#[derive(Clone, Serialize, Default)]
pub struct AbilityCard {
    name: String,
    deck_ix: u32,

    modifiers: Vec<Modifier>,
}

impl AbilityCard {
    fn new(name: &str, deck_ix: u32) -> Self {
        AbilityCard {
            name: name.to_owned(),
            deck_ix,
            modifiers: Vec::new(),
        }
    }
}

pub fn ability_card_deck() -> Vec<AbilityCard> {
    let card1 = AbilityCard::new("Triage", 1);
    let card2 = AbilityCard::new("Focused Mind", 2);
    let card3 = AbilityCard::new("Counsel", 3);

    vec![card1, card2, card3]
}

type Modifier = ();
