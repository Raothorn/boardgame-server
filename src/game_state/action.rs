use core::fmt;
use std::fmt::Display;

use super::{GameState, Update};
use serde::{Deserialize, Serialize};
use serde_json::{Value, json};

mod accept_challenge_result_action;
mod accept_message_action;
mod choose_token_for_deck_action;
mod draw_for_deck_action;
mod handle_event_phase_action;
mod resolve_challenge_action;
mod select_discard_for_galley_action;
mod select_event_option_action;
mod take_ship_action;
mod select_main_action;
mod travel_action;
mod equip_ability_card_action;
mod select_crew_member_action;

#[typetag::serde(tag = "actionType", content = "actionData")]
pub trait Action: fmt::Display {
    fn execute(&self, state: &GameState) -> Update<GameState> {
        Ok(state.clone())
    }
}

pub fn get_action(action_msg_str: &str) -> Box<dyn Action>{
    let msg: Value = serde_json::from_str(action_msg_str).unwrap();

    let result = serde_json::from_value::<Box<dyn Action>>(msg);
    match result {
        Ok(action) => action,
        
        // We want to panic if an action isn't being parsed
        Err(err) => panic!("{}", err),
    }
}


// BASIC ACTIONS

#[derive(Deserialize, Serialize, Clone)]
struct NoAction;

#[typetag::serde(name = "noAction")]
impl Action for NoAction {
    fn execute(&self, gs: &GameState) -> Update<GameState> {
        Ok(gs.to_owned())
    }
}

impl Display for NoAction {
    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
        write!(f, "No action")
    }
}
