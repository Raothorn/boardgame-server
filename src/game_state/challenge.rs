use serde::Serialize;

use super::{skill::Skill, Update, GameState};

#[derive(Clone, Serialize)]
pub struct Challenge {
    pub skill: Skill,
    pub amount: u32,

    #[serde(skip_serializing)]
    pub if_fail: fn(&GameState) -> Update,
    #[serde(skip_serializing)]
    pub if_succeed: fn(&GameState) -> Update
}

impl Default for Challenge {
    fn default() -> Self {
        Self {
            skill: Skill::Craft,
            amount: Default::default(),
            if_fail: |gs| Ok(gs.clone()),
            if_succeed: |gs| Ok(gs.clone())
        }
    }
}
