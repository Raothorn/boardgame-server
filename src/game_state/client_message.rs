use serde::Serialize;

use super::AbilityCard;

#[derive(Clone, Serialize)]
pub enum ClientMessage {
    GainCommandPoints { amount: u32 },
    DrewAbilityCard { card: AbilityCard }
}
