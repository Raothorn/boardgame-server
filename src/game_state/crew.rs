use serde::Serialize;
use serde_with::DisplayFromStr;

use super::skill::Skill;
use std::collections::HashMap;

#[serde_with::serde_as]
#[derive(Serialize, Clone)]
pub struct Crew {
    pub name: String,
    pub fatigue: u8,
    pub damage: u8,
    #[serde_as(as = "HashMap<DisplayFromStr,_>")]
    pub skills: HashMap<Skill, u32>,
}

impl Crew {
    pub fn new(
        name: &str,
        savvy: u32,
        craft: u32,
        strength: u32,
        wits: u32,
        perception: u32,
    ) -> Self {
        let mut skills = HashMap::new();

        skills.insert(Skill::Craft, craft);
        skills.insert(Skill::Savvy, savvy);
        skills.insert(Skill::Strength, strength);
        skills.insert(Skill::Perception, perception);
        skills.insert(Skill::Wits, wits);

        Crew {
            name: name.to_owned(),
            fatigue: 0,
            damage: 0,
            skills,
        }
    }

    pub fn change_fatigue(&mut self, amount: i32) {
        let mut fatigue = i32::from(self.fatigue) + amount;
        // Clamp fatigue to [0, 2]
        if fatigue > 2 {
            fatigue = 2;
        } else if fatigue < 0 {
            fatigue = 0;
        }

        self.fatigue = u8::try_from(fatigue).unwrap_or(0);
    }
}
