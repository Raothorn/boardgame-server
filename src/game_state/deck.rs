use serde::Serialize;

#[derive(Serialize, Clone)]
pub struct Deck<T: Clone> {
    items: Vec<T>,
    discard: Vec<T>,
}

impl<T: Clone> Deck<T> {
    pub fn new(items: Vec<T>) -> Self {
        Deck {
            items,
            discard: Vec::new(),
        }
    }

    pub fn draw(self) -> Result<(Self, T), String> {
        let mut deck = self.clone();
        if deck.items.is_empty() {
            deck.items.append(&mut deck.discard);
            println!("none left");
        }

        deck.items
            .pop()
            .ok_or("No items left in the deck".to_string())
            .map(|item| (deck, item))
    }

    pub fn add_to_discard(&mut self, item: &T) {
        self.discard.push(item.clone());
    }
}
