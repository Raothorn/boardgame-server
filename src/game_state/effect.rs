use std::fmt::Display;

use serde::{Deserialize, Serialize};
use serde_json::json;

use super::{
    action::Action,
    challenge::Challenge,
    crew::Status,
    game_phase::{GamePhase, MainActionSubphase},
    GameState, Resource, Update,
};

#[derive(Clone, Serialize, Debug)]
pub enum Effect {
    TryChallenge(Challenge),
    GainResource(Resource, i32),
    TakeHealthDamage(u32),
    TakeStatus(Status),
    TurnToStory(String),
    ReturnToShip,
}

impl Effect {
    pub fn resolve(&self, state: GameState) -> Update<GameState> {
        match self {
            Effect::TurnToStory(story_ix) => {
                state.map.storybook.story(story_ix).clone().and_then(
                    |story| {
                        state.set_phase(GamePhase::ExplorePhase(
                            story.clone(),
                        ))
                    },
                )
            }

            Effect::ReturnToShip => state.pop_phase().and_then(|g| {
                if let GamePhase::MainActionPhase(
                    Some(MainActionSubphase::Explore),
                    action_ct,
                ) = g.phase()
                {
                    let phase =
                        GamePhase::MainActionPhase(None, action_ct);

                    g.set_phase(phase)
                } else {
                    Err("Wrong phase".to_owned())
                }
            }),

            Effect::GainResource(resource, amount) => {
                state.gain_resource(*resource, *amount)
            }

            Effect::TryChallenge(challenge) => {
                state.challenge(challenge.clone())
            }

            Effect::TakeStatus(status) => {
                let action = TakeStatusAction { status: *status };
                let action_ser = json! ({
                    "actionType": "takeStatusAction",
                    "actionData": action
                })
                .to_string();
                let display = format!("Give {:?} to", status);

                state.push_phase(GamePhase::SelectCrewMemberPhase {
                    crew_ix: None,
                    title: display,
                    callback: action_ser,
                })
            }

            _ => state.push_phase(GamePhase::ResolveEffectPhase(
                self.clone(),
            )),
        }
    }
}

pub fn resolve_effects(
    state: &GameState,
    effects: Vec<Effect>,
) -> Update<GameState> {
    effects.into_iter().fold(Ok(state.clone()), |g, eff| {
        g.and_then(|g| eff.resolve(g))
    })
}

#[derive(Clone, Serialize, Deserialize)]
pub struct TakeStatusAction {
    status: Status,
}

#[typetag::serde(name = "takeStatusAction")]
impl Action for TakeStatusAction {
    fn execute(&self, state: &GameState) -> Update<GameState> {
        // TODO add string to phase
        if let GamePhase::SelectCrewMemberPhase{crew_ix: Some(crew_ix), title: _, callback: _} =
            state.phase()
        {
            state
                .clone()
                .update_crew(crew_ix, |c| c.add_status(self.status))
        } else {
            Err("wrong phase".to_owned())
        }
    }
}

impl Display for TakeStatusAction {
    fn fmt(
        &self,
        f: &mut std::fmt::Formatter<'_>,
    ) -> std::fmt::Result {
        write!(f, "Take Status")
    }
}
