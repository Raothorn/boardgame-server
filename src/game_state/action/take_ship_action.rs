use std::fmt::Display;

use serde::Deserialize;

use super::Action;
use crate::game_state::{
    GamePhase, GameState, ShipActionSubphase, ShipRoom, Update,
};

#[derive(Deserialize)]
pub struct TakeShipAction {
    room: ShipRoom,
    player_ix: usize,
}

impl TakeShipAction {
    fn bridge_action(&self, state: &GameState) -> Update {
        Ok(state.clone())
            .and_then(|g| g.give_command_tokens(self.player_ix, 3))
            .and_then(|g| g.draw_cards(self.player_ix, 1))
            .map(|g| GameState {
                room: ShipRoom::Bridge,
                ..g
            })
            .and_then(|g| g.set_phase(GamePhase::EventPhase(None)))
    }

    fn deck_action(&self, state: &GameState) -> Update {
        let phase = GamePhase::ShipAction(Some(
            ShipActionSubphase::DeckAction {
                search_tokens_drawn: Vec::new(),
            },
        ));
        Ok(state.clone())
            .map(|g| GameState {
                room: ShipRoom::Deck,
                ..g
            })
            .and_then(|g| g.set_phase(phase))
            .map(|g| g.prompt_str("drawForDeckAction"))
    }

    fn galley_action(&self, state: &GameState) -> Update {
        let phase = GamePhase::ShipAction(Some(
            ShipActionSubphase::GalleyAction {
                gain_phase_complete: true,
            },
        ));

        Ok(state.clone())
            .and_then(|g| g.give_command_tokens(self.player_ix, 3))
            .and_then(|g| g.draw_cards(self.player_ix, 2))
            .map(|g| GameState {
                room: ShipRoom::Galley,
                ..g
            })
            .and_then(|g| g.set_phase(GamePhase::EventPhase(None)))
            .map(|g| g.prompt_str("selectDiscardForGalleyAction"))
    }
}

impl Action for TakeShipAction {
    fn execute(&self, state: &GameState) -> Update {
        if let GamePhase::ShipAction(None) = &state.phase() {
            if state.room == self.room {
                Err("You cannot visit the same room twice".to_owned())
            } else {
                match self.room {
                    ShipRoom::Bridge => self.bridge_action(state),
                    ShipRoom::Galley => self.galley_action(state),
                    ShipRoom::Deck => self.deck_action(state),
                    _ => Err(String::from("Not implemented")),
                }
            }
        } else {
            Err("Wrong phase.".to_owned())
        }
    }
}

impl Display for TakeShipAction {
    fn fmt(
        &self,
        f: &mut std::fmt::Formatter<'_>,
    ) -> std::fmt::Result {
        write!(
            f,
            "Take Ship Action:\n Room: {:?}\n PlayerIx: {}",
            self.room, self.player_ix
        )
    }
}

#[cfg(test)]
mod test {
    use super::*;
    use test_case::test_case;

    use ShipActionSubphase as Sas;
    use GamePhase::ShipAction as Sa;
    use GamePhase::EventPhase as Ep;
    use GamePhase::ChallengePhase as Cp;

    #[test]
    fn test_bridge_action() {
        let gs = GameState::init_state();
        let action = TakeShipAction {
            room: ShipRoom::Bridge,
            player_ix: 0,
        };

        let result = action.execute(&gs);
        assert!(result.is_ok());

        insta::with_settings!({sort_maps => true}, {
            // the tests here will force maps to sort
            insta::assert_json_snapshot!(result.unwrap());
        });
    }

    // #[test_case(Sa(Some(Sas::GalleyAction{gain_phase_complete: true})); 
    //             "Current in ship action")]
    #[test_case(Ep(None); "In event phase")]
    // #[test_case(Cp(); "In event phase")]
    // #[test_case(GamePhase)]
    fn test_bridge_action_err_if_wrong_phase(phase: GamePhase) {
        let gs = GameState::init_state();
        let action = TakeShipAction {
            room: ShipRoom::Bridge,
            player_ix: 0,
        };

        let result = action.execute(&gs);
        assert!(result.is_err());
    }
}
