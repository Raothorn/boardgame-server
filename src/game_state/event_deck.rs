use crate::game_state::{
    challenge::Challenge, skill::Skill, Resource,
};

use super::effect::*;
use serde::Serialize;

#[derive(Clone, Serialize, Debug)]
pub struct EventCard {
    pub name: String,
    pub options: Vec<EventOption>,
    // TODO: deck index
    pub deck_index: u32,
}

#[derive(Clone, Serialize, Debug)]
pub struct EventOption {
    pub text: String,
    #[serde(skip_serializing)]
    pub effects: Vec<Effect>,
}

pub fn event_deck() -> Vec<EventCard> {
    vec![
        EventCard {
            name: "Broken Biplane".to_owned(),
            deck_index: 11,
            options: vec![
                EventOption {
                    text: "Help repair the airplane (CRAFT 8)"
                        .to_owned(),
                    effects: vec![Effect::TryChallenge(Challenge {
                        skill: Skill::Craft,
                        amount: 8,
                        if_fail: vec![Effect::TakeHealthDamage(5)],
                        if_succeed: vec![
                            Effect::GainResource(Resource::Coin, 1),
                            Effect::GainResource(Resource::Meat, 1),
                        ],
                    })],
                },
                EventOption {
                    text: "Ignore the plane".to_owned(),
                    // Todo
                    effects: vec![Effect::TakeHealthDamage(1)],
                },
            ],
        },
        // EventCard {
        //     name: todo!(),
        //     options: todo!(),
        // },
    ]
}
