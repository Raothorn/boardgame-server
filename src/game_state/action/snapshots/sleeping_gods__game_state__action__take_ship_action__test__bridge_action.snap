---
source: src/game_state/action/take_ship_action.rs
expression: result.unwrap()
---
{
  "phase": {
    "EventPhase": null
  },
  "players": [
    {
      "command_tokens": 3,
      "hand": [
        {
          "name": "card3"
        }
      ]
    }
  ],
  "crew": [
    {
      "name": "Sofi Odessa",
      "fatigue": 1,
      "skills": {
        "Craft": 2,
        "Savvy": 4
      }
    },
    {
      "name": "Laurant Lapointe",
      "fatigue": 0,
      "skills": {
        "Craft": 2,
        "Savvy": 4
      }
    }
  ],
  "ability_deck": {
    "items": [
      {
        "name": "card1"
      },
      {
        "name": "card2"
      }
    ],
    "discard": []
  },
  "search_token_deck": {
    "items": [
      1,
      2,
      3,
      4,
      5,
      6,
      7
    ],
    "discard": []
  },
  "event_card_deck": {
    "items": [
      {
        "name": "Biting Starfish",
        "options": [
          {
            "text": "Gain 1 Coin"
          },
          {
            "text": "Gain 1 Meat"
          }
        ]
      },
      {
        "name": "Broken Biplane",
        "options": [
          {
            "text": "Help repair the airplane: CRAFT 8"
          }
        ]
      }
    ],
    "discard": []
  },
  "room": "Bridge",
  "resources": {
    "coins": 3,
    "grain": 1,
    "meat": 0
  },
  "prompt": null
}
